gson/src/main/java/com/google/gson/ReflectionAccessFilter.java

```diff
@@ -0,0 +1,187 @@
+ * deserialization is allowed for a class.
+ *
+ * <p>A filter can be useful in multiple scenarios, for example when
+ * upgrading to newer Java versions using the Java Platform Module
+ * System (JPMS). A filter then allows to {@linkplain FilterResult#BLOCK_INACCESSIBLE
+ * prevent making inaccessible members accessible}, even if the used
+ * Java version might still allow illegal access (but logs a warning),
+ * or if {@code java} command line arguments are used to open the inaccessible
+ * packages to other parts of the application. This interface defines some
+ * convenience filters for this task, such as {@link #BLOCK_INACCESSIBLE_JAVA}.
+ *
+ * <p>A filter can also be useful to prevent mixing model classes of a
+ * project with other non-model classes; the filter could
+ * {@linkplain FilterResult#BLOCK_ALL block all reflective access} to
+ * non-model classes.
+ *
+ * <p>A reflection access filter is similar to an {@link ExclusionStrategy}
+ * with the major difference that a filter will cause an exception to be
+ * thrown when access is disallowed while an exclusion strategy just skips
+ * fields and classes.
+ *
+ * @see GsonBuilder#addReflectionAccessFilter(ReflectionAccessFilter)
+ */
+public interface ReflectionAccessFilter {
+  /**
+   * Result of a filter check.
+   */
+  enum FilterResult {
+    /**
+     * Reflection access for the class is allowed.
+     *
+     * <p>Note that this does not affect the Java access checks in any way,
+     * it only permits Gson to try using reflection for a class. The Java
+     * runtime might still deny such access.
+     */
+    ALLOW,
+    /**
+     * The filter is indecisive whether reflection access should be allowed.
+     * The next registered filter will be consulted to get the result. If
+     * there is no next filter, this result acts like {@link #ALLOW}.
+     */
+    INDECISIVE,
+    /**
+     * Blocks reflection access if a member of the class is not accessible
+     * by default and would have to be made accessible. This if unaffected
+     * by any {@code java} command line arguments being used to make packages
+     * accessible.
+     *
+     * <p>Note that this <b>only works for Java 9 and higher</b>, for older
+     * Java versions its functionality will be limited and it might behave like
+     * {@link #ALLOW}. Access checks are only performed as defined by the Java
+     * Language Platform System, restrictions imposed by a {@link SecurityManager}
+     * are not considered.
+     *
+     * <p>This result type is mainly intended to help enforcing the access checks of
+     * the Java Platform Module System. It allows detecting illegal access, even if
+     * the used Java version would only log a warning, or is configured to open
+     * packages for reflection using command line arguments.
+     *
+     * @see AccessibleObject#setAccessible(boolean)
+     * @see AccessibleObject#canAccess(Object)
+     */
+    BLOCK_INACCESSIBLE,
```

Comment:
It appears the used method `AccessibleObject.canAccess` completely ignores whether a module is open for reflective access. Maybe this should be made clearer here. But the question is then how useful this filter result type actually is.